rust/flatbuffers/src/builder.rs

```diff
@@ -754,3 +832,147 @@ impl<'fbb> Default for FlatBufferBuilder<'fbb> {
     }
 }
+
+/// An index that indexes from the reverse of a slice.
+///
+/// Note that while the internal representation is an inde
+/// from the end of a buffer, operations like `Add` and `Sub`
+/// behave like a regular index:
+///
+/// # Examples
+///
+/// ```ignore
+/// let buf = [0, 1, 2, 3, 4, 5];
+/// let idx = ReverseIndex::back() - 2;
+/// assert_eq!(&buf[idx.range_to_end()], &[4, 5]);
+/// assert_eq!(idx.to_forward_index(&buf), 4);
+/// ```
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct ReverseIndex(usize);
+
+impl ReverseIndex {
+    /// Returns an index set to the end.
+    ///
+    /// Note: Indexing this will result in an out of bounds error.
+    pub fn back() -> Self {
+        Self(0)
+    }
+
+    /// Returns a struct equivalent to the range `self..`
+    pub fn range_to_end(self) -> ReverseIndexRangeFrom {
+        ReverseIndexRangeFrom(self)
+    }
+
+    /// Returns a struct equivalent to the range `self..end`
+    pub fn range_to(self, end: ReverseIndex) -> ReverseIndexRange {
+        ReverseIndexRange(self, end)
+    }
+
+    /// Transforms this reverse index into a regular index for the given buffer.
+    pub fn to_forward_index<T>(self, buf: &[T]) -> usize {
+        buf.len() - self.0
+    }
+}
+
+impl Sub<usize> for ReverseIndex {
+    type Output = Self;
+
+    fn sub(self, rhs: usize) -> Self::Output {
+        Self(self.0 + rhs)
+    }
+}
+
+impl SubAssign<usize> for ReverseIndex {
+    fn sub_assign(&mut self, rhs: usize) {
+        *self = *self - rhs;
+    }
+}
+
+impl Add<usize> for ReverseIndex {
+    type Output = Self;
+
+    fn add(self, rhs: usize) -> Self::Output {
+        Self(self.0 - rhs)
+    }
+}
+
+impl AddAssign<usize> for ReverseIndex {
+    fn add_assign(&mut self, rhs: usize) {
+        *self = *self + rhs;
+    }
+}
+impl<T> Index<ReverseIndex> for [T] {
+    type Output = T;
+
+    fn index(&self, index: ReverseIndex) -> &Self::Output {
+        let index = index.to_forward_index(self);
+        &self[index]
+    }
+}
+
+impl<T> IndexMut<ReverseIndex> for [T] {
+    fn index_mut(&mut self, index: ReverseIndex) -> &mut Self::Output {
+        let index = index.to_forward_index(self);
+        &mut self[index]
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct ReverseIndexRangeFrom(ReverseIndex);
+
+impl ReverseIndexRangeFrom {
+    pub fn count(&self) -> usize {
+        self.0 .0
+    }
+}
+
+impl<T> Index<ReverseIndexRangeFrom> for [T] {
+    type Output = [T];
+
+    fn index(&self, index: ReverseIndexRangeFrom) -> &Self::Output {
+        let start = index.0.to_forward_index(self);
+        &self[start..]
+    }
+}
+
+impl<T> IndexMut<ReverseIndexRangeFrom> for [T] {
+    fn index_mut(&mut self, index: ReverseIndexRangeFrom) -> &mut Self::Output {
+        let start = index.0.to_forward_index(self);
+        &mut self[start..]
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct ReverseIndexRange(ReverseIndex, ReverseIndex);
```

Comment:
Why not use `core::ops::Range<ReverseIndex>` and then implement `Index<Range<ReverseIndex>> for [T]`? I think if you do that, then the existing `start..end` syntax will work as expected.

(you'd also need `core::ops::RangeFrom<ReverseIndex>` too for `start..`)

Reply:
Ah, of course. Good ole orphan rules. Well, while verbose, this improves clarity on the `usize` semantics, so I think its still a win.

Final nit: Can you get rid of `ReverseIndexRangeFrom`, since `range_to_end` can return `ReverseIndexRange(head, end())`?

After that, assuming CI is green, are you good with me pushing the merge button?

Reply:
Done. You can hit the button when you're ready